docs/research/blake-256-hash-function.md

```diff
@@ -1,13 +1,302 @@
+| BLAKE-256     |  32-bit   |  <2<sup>64    |  512-bit      |   256-bit     | 128-bit   | 14    |
+
+**<em>Table 1: Characteristics of the BLAKE-256 hash function </em>**
+
+
+### **BLAKE Components**
+
+BLAKE is built on previously analyzed, and reliable components; the hash iterative framework (HAIFA) of Biham and Dunkelman [^4], the local wide-pipe introduced by the LAKE hash function[^6], and the core function inspired by ChaCha function of Bernstein [^5]. The use of well understood building blocks makes BLAKE a hash function that is simple to analyze and implement for cryptanalysts and implementers alike.
+
+
+#### **HAIFA iteration mode**
+
+The construction of a hash output is typically done by splitting the input data referred to as a 'message', into small blocks of fixed length and processed iteratively using a cryptographic compression function. The combination of calls to a compression function for processing the input data is an iteration mode (also referred to as a Domain Extender). The Merkle-Damg˚ard (M-D) construction is the classic iteration mode used by MD5, SHA1, RIPEMD-160, and SHA2 hash functions. M-D splits the variable length input message into equal-sized message blocks **_x_** (i.e., 512 bits) of **_n_** blocks, the last block is padded as required and appended with the length of the message.
+
+![Merkle Damgard Construction](/img/Merkle-Damgard Construction.svg)
+
+**<em>Figure 1: Merkle Damgard Construction</em>**
+
+As shown in Figure 3 above, the output of the compression function **_f_** known as a chaining value **_CV_** and the initial inputs are the Initialization Vector (IV) shown as **_CV<sub>0</sub>_**, with **_h(x)_** being the output after all message blocks are processed. M-D maintains the collision resistance of the compression function however it lacks the pre-image and second pre-image resistance. As a result generic attacks such as the multicollision, the long message second-preimage, herding and length extension are possible on M-D based hash functions [^7] [^8] [^9] [^10] [^11]. 
+
+BLAKE's uses HAIFA, which maintains the valuable properties of the M-D construction and adds to the security and scalability of the transformation. HAIFA is essentially M-D construction but with a mandatory counter (number of bits hashed so far) and an optional salt (random data that used as an additional input).
+
+![HAIFA Construction](/img/HAIFA (BLAKE).svg)
+
+**<em>Figure 2: HAIFA Construction</em>**
+
+This iteration mode solves many of the internal collision problems with M-D construction and provides resistance to herding, long message second pre-image, and length extension attacks as the additional salt and counter simulate distinct compression functions for each data block processed [^4] [^12]. HAIFA is both prefix- and suffix-free and, as a result, is collision-resistant as well as indifferentiable from a random oracle, assuming the underlying compression function is ideal [^1] [^13] [^14]. 
```

Comment:
I think we should say here `both prefix-free and suffix-free`. The hyphen with nothing following seems strange